pkg/controllers/provisioning/scheduling/theoreticalnode.go

```diff
@@ -0,0 +1,137 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scheduling
+
+import (
+	"errors"
+
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
+	"github.com/aws/karpenter/pkg/cloudprovider"
+	"github.com/aws/karpenter/pkg/utils/resources"
+)
+
+// TheoreticalNode is a set of constraints, compatible pods, and possible instance types that could fulfill
+// these constraints. This will be turned into one or more actual node instances within the cluster after bin packing.
+type TheoreticalNode struct {
+	Constraints           *v1alpha5.Constraints
```

Comment:
Thoughts on inlining this type?

Reply:
Yes, but they're all bad 😆 To me it doesn't scream out to be inlined and would expose the Constraints methods on the Node which I think reduces readability:

```go
var n Node
n.Constraints = n.Tighten(pod)
```